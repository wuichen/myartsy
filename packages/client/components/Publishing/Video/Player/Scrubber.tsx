import React, { Component } from "react"
import styled, { StyledFunction } from "styled-components"

interface Props extends React.HTMLProps<HTMLDivElement> {
  duration: number
  currentTime: number
  isPlaying: boolean
  pause: () => void
  play: () => void
  seekTo: (v) => void
}

interface InputProps extends React.HTMLProps<HTMLInputElement> {
  duration: number
  currentTime: number
}

interface State {
  isScrubbing: boolean
  isPlayingOnMouseDown: boolean
}

export class Scrubber extends Component<Props, State> {
  state = {
    isScrubbing: false,
    isPlayingOnMouseDown: false,
  }

  shouldComponentUpdate(nextProps) {
    return (
      this.props.currentTime !== nextProps.currentTime ||
      this.props.duration !== nextProps.duration
    )
  }

  handleMouseDown = () => {
    this.setState({
      isPlayingOnMouseDown: this.props.isPlaying,
    })
    this.props.pause()
  }

  handleMouseUp = e => {
    if (!this.state.isScrubbing) {
      this.props.seekTo(e.target.value)
    }
    if (this.state.isPlayingOnMouseDown) {
      this.props.play()
    }
    this.setState({ isScrubbing: false })
  }

  handleChange = e => {
    this.setState({ isScrubbing: true })
    this.props.seekTo(e.target.value)
  }

  render() {
    const { duration, currentTime } = this.props

    return (
      <ScrubberInput
        type="range"
        step="any"
        max={duration && duration.toFixed(4)}
        value={currentTime}
        duration={duration}
        currentTime={currentTime}
        onMouseDown={this.handleMouseDown}
        onMouseUp={this.handleMouseUp}
        onChange={this.handleChange}
      />
    )
  }
}

const Input: StyledFunction<InputProps & React.HTMLProps<HTMLInputElement>> =
  styled.input

const ScrubberInput = Input`
  width: 100%;
  background-size: ${props => (props.currentTime * 100) / props.duration}% 100%;
  height: 18px;
  margin-top: 15px;
  appearance: none;
  background: transparent;
  ::-webkit-slider-thumb {
    appearance: none;
  }
  ::-ms-track {
    width: 100%;
    height: 2px;
    cursor: pointer;
    background: transparent;
    border: 0px;
    color: transparent;
  }
  ::-webkit-slider-runnable-track {
    height: 2px;
    cursor: pointer;
    animate: 0.2s;
    background-color: white;
    border: 0px;
  }
  ::-moz-range-track {
    height: 2px;
    cursor: pointer;
    animate: 0.2s;
    background-color: white;
    border: 0px;
  }
  ::-webkit-slider-thumb {
    color: white;
    border: 0px;
    height: 12px;
    width: 12px;
    border-radius: 50px;
    background: white;
    cursor: pointer;
    margin-top: -5px;
    appearance: none;
  }
  ::-moz-range-thumb {
    color: white;
    box-shadow: 0px 0px 0px white;
    border: 0px;
    height: 12px;
    width: 12px;
    border-radius: 50px;
    background: white;
    cursor: pointer;
    margin-top: -5px;
    appearance: none;
  }
  ::-ms-fill-lower {
    background: white;
    border: 0px;
  }
  ::-ms-fill-upper {
    background: white;
    border: 0px;
  }
  ::-ms-thumb {
    color: white;
    box-shadow: 0px 0px 0px white;
    border: 0px;
    height: 12px;
    width: 12px;
    border-radius: 50px;
    background: white;
    cursor: pointer;
    margin-top: -5px;
    appearance: none;
  }
  &:focus::-webkit-slider-runnable-track {
    background: white;
  }
  &:focus {
    outline: none;
  }
`
