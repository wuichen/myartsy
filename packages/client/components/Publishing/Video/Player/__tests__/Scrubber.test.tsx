import { mount } from "enzyme"
import "jest-styled-components"
import React from "react"
import renderer from "react-test-renderer"
import { Scrubber } from "../Scrubber"

describe("Scrubber", () => {
  it("matches the snapshot", () => {
    const scrubber = renderer
      .create(
        <Scrubber
          isPlaying={false}
          duration={5000}
          currentTime={0}
          pause={jest.fn()}
          play={jest.fn()}
          seekTo={jest.fn()}
        />
      )
      .toJSON()
    expect(scrubber).toMatchSnapshot()
  })

  it("handles #onMouseDown", () => {
    const pause = jest.fn()
    const component = mount(
      <Scrubber
        isPlaying={false}
        duration={5000}
        currentTime={0}
        pause={pause}
        play={jest.fn()}
        seekTo={jest.fn()}
      />
    )
    component.simulate("mouseDown")
    expect(pause).toBeCalled()
    expect(component.state("isPlayingOnMouseDown")).toBe(false)
  })

  it("sets isPlayingOnMouseDown if the video is playing on click", () => {
    const pause = jest.fn()
    const component = mount(
      <Scrubber
        isPlaying
        duration={5000}
        currentTime={0}
        pause={pause}
        play={jest.fn()}
        seekTo={jest.fn()}
      />
    )
    component.simulate("mouseDown")
    expect(pause).toBeCalled()
    expect(component.state("isPlayingOnMouseDown")).toBe(true)
  })

  it("resumes video and seeksTo on #onMouseUp if it was playing", () => {
    const play = jest.fn()
    const seekTo = jest.fn()
    const component = mount(
      <Scrubber
        isPlaying
        duration={5000}
        currentTime={0}
        pause={jest.fn()}
        play={play}
        seekTo={seekTo}
      />
    )
    component.setState({ isPlayingOnMouseDown: true })
    component.simulate("mouseup", { target: { value: 1000 } })
    expect(seekTo.mock.calls[0][0]).toBe(1000)
    expect(play).toBeCalled()
  })

  it("does not resume or seekTo #onMouseUp", () => {
    const play = jest.fn()
    const seekTo = jest.fn()
    const component = mount(
      <Scrubber
        isPlaying
        duration={5000}
        currentTime={0}
        pause={jest.fn()}
        play={play}
        seekTo={seekTo}
      />
    )
    component.setState({ isScrubbing: true })
    component.simulate("mouseup", { target: { value: 1000 } })
    expect(play).toHaveBeenCalledTimes(0)
    expect(seekTo).toHaveBeenCalledTimes(0)
  })

  it("allows single click scrubbing with #handleChange", () => {
    const play = jest.fn()
    const seekTo = jest.fn()
    const component = mount(
      <Scrubber
        isPlaying
        duration={5000}
        currentTime={0}
        pause={jest.fn()}
        play={play}
        seekTo={seekTo}
      />
    )
    component.simulate("change", { target: { value: 1000 } })
    expect(seekTo.mock.calls[0][0]).toBe(1000)
  })
})
