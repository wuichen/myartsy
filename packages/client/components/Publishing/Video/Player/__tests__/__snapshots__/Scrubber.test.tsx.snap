// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scrubber matches the snapshot 1`] = `
.c0 {
  width: 100%;
  background-size: 0% 100%;
  height: 18px;
  margin-top: 15px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: transparent;
}

.c0::-webkit-slider-thumb {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c0::-ms-track {
  width: 100%;
  height: 2px;
  cursor: pointer;
  background: transparent;
  border: 0px;
  color: transparent;
}

.c0::-webkit-slider-runnable-track {
  height: 2px;
  cursor: pointer;
  -webkit-animate: 0.2s;
  animate: 0.2s;
  background-color: white;
  border: 0px;
}

.c0::-moz-range-track {
  height: 2px;
  cursor: pointer;
  -webkit-animate: 0.2s;
  animate: 0.2s;
  background-color: white;
  border: 0px;
}

.c0::-webkit-slider-thumb {
  color: white;
  border: 0px;
  height: 12px;
  width: 12px;
  border-radius: 50px;
  background: white;
  cursor: pointer;
  margin-top: -5px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c0::-moz-range-thumb {
  color: white;
  box-shadow: 0px 0px 0px white;
  border: 0px;
  height: 12px;
  width: 12px;
  border-radius: 50px;
  background: white;
  cursor: pointer;
  margin-top: -5px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c0::-ms-fill-lower {
  background: white;
  border: 0px;
}

.c0::-ms-fill-upper {
  background: white;
  border: 0px;
}

.c0::-ms-thumb {
  color: white;
  box-shadow: 0px 0px 0px white;
  border: 0px;
  height: 12px;
  width: 12px;
  border-radius: 50px;
  background: white;
  cursor: pointer;
  margin-top: -5px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c0:focus::-webkit-slider-runnable-track {
  background: white;
}

.c0:focus {
  outline: none;
}

<input
  className="c0"
  max="5000.0000"
  onChange={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  step="any"
  type="range"
  value={0}
/>
`;
